### PR TITLE
[Security] Idle sessions expiration.

### PR DESCRIPTION
**Pull request by ajgarlag**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR | symfony/symfony-docs#4576 |

This new firewall allows you to block long running sessions.

When the session cookie is valid until the browser is closed (`cookie_lifetime = 0`), but the user left the browser open for a long time, the session will be invalidated if the `max_idle_time` is exceeded. The `max_idle_time` value should be lesser or equal than the one from `session.gc_maxlifetime`.

If an `expiration_url` is set, the user agent will be redirected to this URL after the session invalidation. If this parameter is not set, a `SessionExpiredException` will be thrown.

### CODE REVIEW

File: src/Symfony/Bundle/SecurityBundle/Tests/Functional/SessionExpirationTest.php
Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/SessionExpirationTest.php]: we generally don't put ``@author`` tags on test files

File: src/Symfony/Component/Security/Core/Exception/SessionExpiredException.php
Comment (stof) [inline src/Symfony/Component/Security/Core/Exception/SessionExpiredException.php]: this key needs to be added in the translation files of the Security component (at least the English file which is the reference, and then any other language for which you are able to translate it yourself)

File: src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php
Comment (stof) [inline src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php]: please don't trigger warnings. Either it is an error and it should be an exception, or it is not an error but only a notice to improve things and it should be logged rather than triggering a warning (warnings are a pain to handle because they are not catchable easily)

File: src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php
Comment (stof) [inline src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php]: please separate the ``@param`` and ``@throws`` with an empty phpdoc line

File: src/Symfony/Component/Security/Http/Tests/Firewall/SessionExpirationListenerTest.php
Comment (stof) [inline src/Symfony/Component/Security/Http/Tests/Firewall/SessionExpirationListenerTest.php]: this is wrong. the code returns either a SessionInterface instance or ``null``. It never returns ``false``.

File: src/Symfony/Component/Security/Http/Tests/Firewall/SessionExpirationListenerTest.php
Comment (stof) [inline src/Symfony/Component/Security/Http/Tests/Firewall/SessionExpirationListenerTest.php]: use ``getMockBuilder``. It will be much more readable than passing all the arguments to ``getMock`` (I don't know what the order of the getMock arguments is, and I doubt lots of people are knowing it without checking the doc given the complexity of the signature, which is precisely why there is a mock builder)

**Comment by ajgarlag** [comment]
@stof Thanks for your review. I think all errors you found were fixed.

File: src/Symfony/Bundle/SecurityBundle/Resources/config/security_listeners.xml
Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/Resources/config/security_listeners.xml]: I would call it ``session_expiration_listener`` (separating all words with underscores, not only half of them)

**Comment by fabpot** [comment]
ping @symfony/deciders What about this feature? Is it something we want in core?

File: src/Symfony/Bundle/SecurityBundle/Tests/Functional/app/SessionExpiration/config.yml
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/app/SessionExpiration/config.yml]: What do we need this firewall for?

File: src/Symfony/Bundle/SecurityBundle/Tests/Functional/app/SessionExpiration/config.yml
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/app/SessionExpiration/config.yml]: missing newline

File: src/Symfony/Bundle/SecurityBundle/Tests/Functional/app/SessionExpiration/routing.yml
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/app/SessionExpiration/routing.yml]: blank line can be removed

File: src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php
Comment (xabbuh) [inline src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php]: I think you can inline that check.

File: src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php
Comment (xabbuh) [inline src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php]: no need for the extra method imo

File: src/Symfony/Bundle/SecurityBundle/Resources/config/security_listeners.xml
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/Resources/config/security_listeners.xml]: You now also need to update the required version of the `symfony/security-http` component in the SecurityBundle.

File: src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php
Comment (ajgarlag) [inline src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php]: I think is cleaner with the extra method.

File: src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php
Comment (ajgarlag) [inline src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php]: I think it is cleaner with the extra method.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: I guess using `null` here and determine the ini setting value at runtime then would be safer.

**Review by xabbuh — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/SecurityBundle/Resources/config/security_listeners.xml
Comment (xabbuh) [inline src/Symfony/Bundle/SecurityBundle/Resources/config/security_listeners.xml]: Modifying the existing requirement looks okay to me.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php
Comment (ajgarlag) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: I'm not sure about it. If this ini option has different values for web and cli environments, it will be problematic:
if we set it at runtime, this value can vary between http requests and console commands; but if we set it at config time, the value will be read only once when the container is compiled, and the maybe the value is not the preferred one.

Anyway, ICYMI the `E_USER_WARNING` was removed and now the behavior is to log the problem with `WARN` level.

Maybe we should remove any default value and require the user to provide it.

Comment (ajgarlag) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: @xabbuh I'm waiting for your final decision here:

* Set the option at config time
* Set the value at runtime setting the option to null at config time.
* Make the option required

I'll implement your preferred option and the PR will be ready for merge.

Comment (ajgarlag) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: ping @xabbuh

**Comment by weaverryan** [comment]
What's the advantage of this over using `session.gc_maxlifetime` and allowing the session to expire naturally? Is it that you can expire your login, but before expiring your session? Or is it because you want to be able to control this on a firewall-by-firewall basis? 

I just want to make sure I understand the need for this. :)

**Comment by ajgarlag** [comment]
The `session.gc_maxlifetime` param only indicates to the garbage collector that sessions older than that can be safely removed. But the garbage collector is started with a probability of `session.gc_probability` divided by `session.gc_divisor`, so you couldn't really control it until PHP 7.1 (this PR is 3 years old) where there is a new `session_gc` function that you can call to launch the garbage collection of session data.

Another problem with `session.gc_maxlifetime` is that if you use the native session save handler, you must use a filesystem that modifies the access time of the session file. Otherwise, the max lifetime will be computed from the session start, and not from last access.

Another advantage of this PR is that when an expired session is detected, you can control how your app should react, catching the `SessionExpiredException`. With the garbage collection mechanism, when it is launched, it removes data from all sessions, not only for the one active in that request, so if an user which session data have been removed returns to your app his session is gone, you cannot identify him an provide an user friendly message like *Your session have expired due inactivity*.

You can read a more detailed explanation in https://stackoverflow.com/questions/520237/how-do-i-expire-a-php-session-after-30-minutes

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php
Comment (weaverryan) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: we should make the option required (if the `session_expiration` key is present)

File: src/Symfony/Bundle/SecurityBundle/Tests/Functional/Bundle/SimpleLoginBundle/SimpleLoginBundle.php
Comment (weaverryan) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/Bundle/SimpleLoginBundle/SimpleLoginBundle.php]: Could we just use the `FormLoginBundle` directly in the tests? Then we could remove this file and the new `after_login.html.twig`. That bundle already gives us a functional `/profile` URL, which is all we really need I think.

File: src/Symfony/Bundle/SecurityBundle/Tests/Functional/SessionExpirationTest.php
Comment (weaverryan) [inline src/Symfony/Bundle/SecurityBundle/Tests/Functional/SessionExpirationTest.php]: `// Wait` (add a space)

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php
Comment (weaverryan) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: What about `expiration_path`? This seems more consistent with the keys under `form_login`, for example. And this could also be a URL, correct? It looks like you set it up that way :)

File: src/Symfony/Bundle/SecurityBundle/composer.json
Comment (weaverryan) [inline src/Symfony/Bundle/SecurityBundle/composer.json]: Can be removed - on the `3.4` branch, this is already updated to `"symfony/security": "~3.4|~4.0",`

File: src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php
Comment (weaverryan) [inline src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php]: nit pick, but use `'` around the string and then `"%s"` inside. We almost always wrap strings in single quotes (sometimes this isn't followed, but usually when we need `"\n"`)

File: src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php
Comment (weaverryan) [inline src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php]: Ok, so this is the trickiest part of this PR. You're effectively imitating what `LogoutListener` does + some of its logout handlers (e.g. `SessionLogoutHandler`). So then (to play devil's advocate), why not actually inject the logout handlers and use them here? We *are* logging the user out... so shouldn't those handlers be called? And to go further, couldn't we use the normal `DefaultLogoutSuccessHandler` to figure out where to take the user next?

The disadvantage to this is that you couldn't do something different based on a normal user logout versus a session expiration logout. But, is that a problem? And if it is, could we dispatch a special event (e.g. `security.session_expired`) in this class, *before* calling the logout handlers?

File: src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php
Comment (weaverryan) [reply src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php]: I agree with @xabbuh in this case

**Review by weaverryan — CHANGES_REQUESTED** [review]
This is a sound PR imo. And so, I'd like to see it merged :). I just sent you some feedback @ajgarlag. Sorry for the long delays - sometimes security stuff (due to complexity, sensitivity) is one that is the most difficult to review. So, thanks for your patience. I only see one potential blocker, and it's about the logout handlers inside the listener.

File: src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php
Comment (ajgarlag) [reply src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php]: After 3 years, my vision is slightly different. I'll remove the `$tokenStorage->setToken(null)` so this listener will only detect the session expiration and optionally (enabled by default) will invalidate the session.

For me, the logout is an interactive user action and different from the session expiration case.
If you need to force the logout you could disable the session invalidation and redirect the user to the logout path. What do you think?

Comment (weaverryan) [reply src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php]: Hmm... so then the purpose of this listener would be to *only* invalidate the session after a certain amount of idle time. If that's the case... then this is kind of unrelated to security, right? This could just be a general framework feature of the session: a framework-level way to invalidate sessions that are idle. Wdyt?

Comment (ajgarlag) [reply src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php]: @weaverryan It took me some time to review this, as the history behind this PR is too old. It is the result of an split of #12009, which was a reimplementation of #786, into #12807(this PR) and #12810, so I decided to write a recap

We have to implement 2 different things here: first, a feature to detect idle sessions, and last, to provide a default behavior once the expired session is detected.

**Detecting idle sessions**

It is now implemented as a firewall listener, so it can be enabled or disabled for different firewalls in the same app, but if the session is invalidated inside one firewall, it will logout from any firewall that uses the `ContextListener` to restore the token.

If we decide that this feature is not related to security, but to the framework, we can implement it as a `kernel.request` event listener that, IMO, should be executed with higher priority than the firewall one, so the user is not authenticated from an expired session. I'm not sure which component (HttpKernel?) should provide this listener.

Anyway, once the expired session is detected, I think that a new event should be dispatched (currently not implemented). Your proposed name `security.session_expired` it's only an option if we decide that this feature is a security concern.

**Default behavior on expired sessions**

For me, the best option here is to invalidate the session (should we allow not to override this?), and redirect the user to a given path that will return an accurate response depending on the given request (html, json, etc...) or to throw a `SessionExpiredException` in no path is defined. If we implement this feature as a kernel event listener, this behavior should be hooked to the new `session_expired` event.

In a previous comment you proposed to force an user logout calling the `LogoutSuccessHandler` and all defined `LogoutHandlers`, but I see two problems here:
  1. In the app there could be more than one firewall, so; which logout should be executed when an expired session is detected if we decide that this option should not be implemented as a firewall listener, but as a kernel listener?
  2. If we decide to execute the logout from the current firewall, (in this case the expired session listener should be a firewall listener, or has to be executed after the firewall listener), what should be done if there is no logout listener defined for the current path?

File: src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php
Comment (oleg-andreyev) [reply src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php]: `getLastUsed` will return `null` if session is not started

File: src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php
Comment (weaverryan) [reply src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php]: Hey @ajgarlag!

Sorry for the slow reply again :p.

After reading your reply, it seems to me that this PR indeed has nothing to do with security. We're simply implementing a framework-level "session timeout" feature. This is necessary because there's no reliable way to do this with php.ini configuration. If there were, then we would just have the user do that.

With that in mind, this 100% seems like "session" configuration. And so, it would be configured under the `framework.session` configuration. And I don't think ANY other configuration is necessary - no need to have `expiration_url`. Nope, we would simply invalidate the user's session and that's it (e.g. allow the request to continue).

IF we want to allow the user to hook into this further, then dispatching the event we discussed may make sense. I WOULD do that, then we can see what others think.

To summarize (and tell me if you disagree):

1) This is not related to security at all - implement it under the `framework.session` configuration. The listener would be configured in FrameworkBundle, but the listener/subscriber class itself would live in the HttpKernel component.

2) When a session "expires", simply invalidate the session and do nothing else. Well, let's also dispatch a new event, and we can see if people think it's a good idea.

**Comment by nicolas-grekas** [comment]
I'm chiming in a bit randomly so my comment might be totally wrong.
What's the difference with using a remember me cookie here?
Right now, all session storages update the last-access timestamp and enforce gc_maxlifetime.
In order to detect idle sessions, one can set gc_maxlifetime to the idle time they want + rely on a remember-me cookie to reboot a fresh session after the idle timeout expired, providing adjusted role management + privilege upgrade path, using only existing mechanisms from core, isn't it?

**Comment by fabpot** [comment]
Looks like @nicolas-grekas is right. Closing.

**Pull request closed without merging**
